rusk/src/lib/services/version.rs

```diff
@@ -0,0 +1,156 @@
+        RuskVersionMiddleware { inner: service }
+    }
+}
+
+/// This middleware adds `x-rusk-version` to response headers
+/// for any server response, be it the result of a sucessful
+/// request or not.
+#[derive(Debug, Clone)]
+pub struct RuskVersionMiddleware<S> {
+    inner: S,
+}
+
+impl<S> Service<hyper::Request<Body>> for RuskVersionMiddleware<S>
+where
+    S: Service<hyper::Request<Body>, Response = hyper::Response<BoxBody>>
+        + Clone
+        + Send
+        + 'static,
+    S::Future: Send + 'static,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = futures::future::BoxFuture<
+        'static,
+        Result<Self::Response, Self::Error>,
+    >;
+
+    fn poll_ready(
+        &mut self,
+        cx: &mut Context<'_>,
+    ) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: hyper::Request<Body>) -> Self::Future {
+        // This is necessary because tonic internally uses
+        // `tower::buffer::Buffer`. See https://github.com/tower-rs/tower/issues/547#issuecomment-767629149
+        // for details on why this is necessary
+        let clone = self.inner.clone();
+        let mut inner = std::mem::replace(&mut self.inner, clone);
+
+        Box::pin(async move {
+            let mut response = inner.call(req).await?;
+            let headers = response.headers_mut();
+            headers.append("x-rusk-version", HeaderValue::from_static(VERSION));
+            Ok(response)
+        })
+    }
+}
+
+/// Checks incoming requests for compatibility with running
+/// Rusk version using an Interceptor.
+#[derive(Clone)]
+pub struct CompatibilityInterceptor;
+
+impl Interceptor for CompatibilityInterceptor {
+    fn call(
+        &mut self,
+        request: tonic::Request<()>,
+    ) -> Result<tonic::Request<()>, Status> {
+        // attempt to extract `x-rusk-version` header metadata
+        let metadata = request.metadata();
+        match metadata.get("x-rusk-version") {
+            Some(header_v) => {
+                // extract a string value
+                let mut client_version = "unknown";
+                let is_compat = match header_v.to_str() {
+                    Ok(req_v) => {
+                        // check for compatibility
+                        client_version = req_v;
+                        is_compatible(req_v, VERSION)
+                    }
+                    Err(_) => false,
+                };
+                if !is_compat {
+                    return Err(Status::failed_precondition(
+                        format!("Requested rusk version is not supported. Expected {} but got {}!", VERSION, client_version),
+                    ));
+                }
+            }
+            None => {
+                return Err(Status::unavailable(
+                    "Missing \"x-rusk-version\" header, please update client!",
+                ))
+            }
+        }
+        Ok(request)
```

Comment:
This part could be probably simplified but for the time being let's leave as is.